guides/keras_cv/fine_tune_via_textual_inversion.py

```diff
@@ -0,0 +1,691 @@
+!pip install -q luketils
+!pip install -q tensorflow==2.11.0
+"""
+import keras_cv
+from keras_cv import layers as cv_layers
+import tensorflow as tf
+from tensorflow import keras
+from luketils import visualization
+import numpy as np
+import random
+
+stable_diffusion = keras_cv.models.StableDiffusion()
+
+"""
+## Assembling a text-image pair dataset
+
+In order to train the embedding of our new token, we first must assemble a dataset
+consisting of text-image pairs.
+Each sample from the dataset must contain an image of the concept we are teaching
+StableDiffusion, as well as a caption accurately representing the content of the image.
+
+In this tutorial, we will teach StableDiffusion the concept of Luke and Ian's GitHub
+avatar's:
+
+![gh-avatars](https://i.imgur.com/WyEHDIR.jpg)
+
+First, lets construct an image dataset of cat dolls:
+"""
+
+
+def assemble_image_dataset(urls):
+    # fetch all remote files
+    files = [tf.keras.utils.get_file(origin=url) for url in urls]
+
+    # Resize images
+    resize = keras.layers.Resizing(height=512, width=512, crop_to_aspect_ratio=True)
+    images = [keras.utils.load_img(img) for img in files]
+    images = [keras.utils.img_to_array(img) for img in images]
+    images = np.array([resize(img) for img in images])
+
+    # plot images
+    visualization.plot_gallery(
+        images, value_range=(0, 255), rows=1, cols=3, scale=3, show=True
+    ),
+
+    # pipe into tf.data.Dataset
+    images = images / 127.5 - 1
+    image_dataset = tf.data.Dataset.from_tensor_slices(images)
+
+    # shuffle and introduce random noise
+    image_dataset = image_dataset.shuffle(50, reshuffle_each_iteration=True)
+    image_dataset = image_dataset.map(
+        cv_layers.RandomCropAndResize(
+            target_size=(512, 512),
+            crop_area_factor=(0.8, 1.0),
+            aspect_ratio_factor=(1.0, 1.0),
+        ),
+        num_parallel_calls=tf.data.AUTOTUNE,
+    )
+    image_dataset = image_dataset.map(
+        cv_layers.RandomFlip(mode="horizontal"),
+        num_parallel_calls=tf.data.AUTOTUNE,
+    )
+    return image_dataset.repeat()
+
+
+"""
+Next, we will need to assemble a text dataset:
+"""
+
+MAX_PROMPT_LENGTH = 77
+placeholder_token = "<my-funny-cat-token>"
+
+
+def pad_embedding(embedding):
+    return embedding + (
+        [stable_diffusion.tokenizer.end_of_text] * (MAX_PROMPT_LENGTH - len(embedding))
+    )
+
+
+stable_diffusion.tokenizer.add_tokens(placeholder_token)
+
+
+def assemble_text_dataset(prompts, repeats=5):
+    prompts = [prompt.format(placeholder_token) for prompt in prompts]
+    embeddings = [stable_diffusion.tokenizer.encode(prompt) for prompt in prompts]
+    embeddings = [np.array(pad_embedding(embedding)) for embedding in embeddings]
+    text_dataset = tf.data.Dataset.from_tensor_slices(embeddings)
+    text_dataset = text_dataset.shuffle(100, reshuffle_each_iteration=True)
+    return text_dataset.repeat(repeats)
```

Comment:
The fact that the first dataset was indefinitely repeated but this one is not is confusing